jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/ExternalCheckpointManager.java

```diff
@@ -0,0 +1,116 @@
+package ml.dmlc.xgboost4j.java;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+public class ExternalCheckpointManager {
+
+  private Log logger = LogFactory.getLog("ExternalCheckpointManager");
+  private String modelSuffix = ".model";
+  private String checkpointPath;
+  private FileSystem fs;
+
+  public ExternalCheckpointManager(String checkpointPath, FileSystem fs) {
+    this.checkpointPath = checkpointPath;
+    this.fs = fs;
+  }
+
+  private String getPath(int version) {
+    return checkpointPath + "/" + version + modelSuffix;
+  }
+
+  private List<Integer> getExistingVersions() throws IOException {
+    if (checkpointPath == null || checkpointPath.isEmpty() ||
+            !fs.exists(new Path(checkpointPath))) {
+      return new ArrayList<>();
+    } else {
+      return Arrays.stream(fs.listStatus(new Path(checkpointPath)))
+              .map(path -> path.getPath().getName())
+              .filter(fileName -> fileName.endsWith(modelSuffix))
+              .map(fileName -> Integer.valueOf(
+                      fileName.substring(0, fileName.length() - modelSuffix.length())))
+              .collect(Collectors.toList());
+    }
+  }
+
+  public void cleanPath() throws IOException {
+    if (checkpointPath != null) {
+      fs.delete(new Path(checkpointPath), true);
+    }
+  }
+
+  public Booster loadCheckpointAsBooster() throws IOException, XGBoostError {
+    List<Integer> versions = getExistingVersions();
+    if (versions.size() > 0) {
+      int latestVersion = versions.stream().max(Comparator.comparing(Integer::valueOf)).get();
+      String checkpointPath = getPath(latestVersion);
+      InputStream in = fs.open(new Path(checkpointPath));
+      logger.info("loaded checkpoint from " + checkpointPath);
+      Booster booster = XGBoost.loadModel(in);
```

Comment:
I have a RFC for distinguishing `model` and `checkpoint`: https://github.com/dmlc/xgboost/issues/4855 .  Which is implemented in #4732  with simple documentation.  The basic idea is when you perform `SaveModel` only the trees and objective are saved, while `SaveCheckPoints` also saves all hyper-parameters like `eta, max_depth, tree_method` etc that's guaranteed to carry out the previous training.  By carrying out previous training, I mean:
``` 
 for i in range(4):
    bst.update_one_iter()

  bst.save()
  bst.load()

  for i in range(4):
    bst.update_one_iter()
```
Should equals to following when prediction cache is not involved:
```
  for i in range(8):
    bst.updater_one_iter()
```
It seems `XGBoosterLoadRabitCheckpoint` is more appropriate here.

Reply:
hmmm....in JVM part, we actually not that care about it, as the training parameters are corrected by what users passed in training API


e.g. 

1. we load a booster with default parameter at https://github.com/dmlc/xgboost/pull/5082/files#diff-095250777d030169ae83818b37ab44b1R530

2. we continue training with https://github.com/dmlc/xgboost/pull/5082/files#diff-095250777d030169ae83818b37ab44b1R543-R544

3. in buildDistributedBoosters https://github.com/dmlc/xgboost/pull/5082/files#diff-095250777d030169ae83818b37ab44b1R442-R443 we will call Java API like https://github.com/dmlc/xgboost/pull/5082/files#diff-095250777d030169ae83818b37ab44b1R354-R357

4. before we start  training, we'll set params https://github.com/dmlc/xgboost/pull/5082/files#diff-314932ca1cabd6bd148e8f8bd85b8947R190

Reply:
@CodingCat Can we consider replacing all these with the JSON implementation after I manage to merge that PR? Maybe next release after I tested it enough on dask side?  You can load/save only hyper-parameters in that PR.

Reply:
@CodingCat I provided a sample output of `XGBoosterSaveJsonParameters` in #4732 's document.  Pasted below.  I believe it's more robust than language binding implementation as it actually walks through all parameters.  Also with it we can build some consistency between language bindings.


``` json
{
      "Learner": {
        "generic_parameter": {
          "enable_experimental_json_serialization": "0",
          "gpu_id": "0",
          "gpu_page_size": "0",
          "n_jobs": "0",
          "random_state": "0",
          "seed": "0",
          "seed_per_iteration": "0"
        },
        "gradient_booster": {
          "gbtree_train_param": {
            "num_parallel_tree": "1",
            "predictor": "gpu_predictor",
            "process_type": "default",
            "tree_method": "gpu_hist",
            "updater": "grow_gpu_hist",
            "updater_seq": "grow_gpu_hist"
          },
          "name": "gbtree",
          "updater": {
            "grow_gpu_hist": {
              "gpu_hist_train_param": {
                "debug_synchronize": "0",
                "gpu_batch_nrows": "0",
                "single_precision_histogram": "0"
              },
              "train_param": {
                "alpha": "0",
                "cache_opt": "1",
                "colsample_bylevel": "1",
                "colsample_bynode": "1",
                "colsample_bytree": "1",
                "default_direction": "learn",
                "enable_feature_grouping": "0",
                "eta": "0.300000012",
                "gamma": "0",
                "grow_policy": "depthwise",
                "interaction_constraints": "",
                "lambda": "1",
                "learning_rate": "0.300000012",
                "max_bin": "256",
                "max_conflict_rate": "0",
                "max_delta_step": "0",
                "max_depth": "6",
                "max_leaves": "0",
                "max_search_group": "100",
                "refresh_leaf": "1",
                "sketch_eps": "0.0299999993",
                "sketch_ratio": "2",
                "subsample": "1"
              }
            }
          }
        },
        "learner_train_param": {
          "booster": "gbtree",
          "disable_default_eval_metric": "0",
          "dsplit": "auto",
          "objective": "reg:squarederror"
        },
        "metrics": [],
        "objective": {
          "name": "reg:squarederror",
          "reg_loss_param": {
            "scale_pos_weight": "1"
          }
        }
      },
      "version": [1, 0, 0]
    }
```

Reply:
sure